logs_sdk/lib/opentelemetry/sdk/logs/logger_provider.rb

```diff
@@ -0,0 +1,127 @@
+# frozen_string_literal: true
+
+# Copyright The OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  module SDK
+    module Logs
+      # The SDK implementation of OpenTelemetry::Logs::LoggerProvider.
+      class LoggerProvider < OpenTelemetry::Logs::LoggerProvider
+        attr_reader :resource, :log_record_processors
+
+        UNEXPECTED_ERROR_MESSAGE = 'unexpected error in ' \
+          'OpenTelemetry::SDK::Logs::LoggerProvider#%s'
+        # Returns a new LoggerProvider instance.
+        #
+        # @param [optional Resource] resource The resource to associate with
+        #   new LogRecords created by {Logger}s created by this LoggerProvider.
+        # @param [optional Array] log_record_processors The
+        #   {LogRecordProcessor}s to associate with this LoggerProvider.
+        #
+        # @return [OpenTelemetry::SDK::Logs::LoggerProvider]
+        def initialize(
+          resource: OpenTelemetry::SDK::Resources::Resource.create,
+          log_record_processors: []
+        )
+          @log_record_processors = log_record_processors
+          @mutex = Mutex.new
+          @resource = resource
+          @stopped = false
+        end
+
+        # Creates an {OpenTelemetry::SDK::Logs::Logger} instance.
+        #
+        # @param [optional String] name Instrumentation package name
+        # @param [optional String] version Instrumentation package version
+        #
+        # @return [OpenTelemetry::SDK::Logs::Logger]
+        def logger(name = nil, version = nil)
+          name ||= ''
+          version ||= ''
+
+          OpenTelemetry.logger.warn('LoggerProvider#logger called without providing a logger name.') if name.empty?
+
+          OpenTelemetry::SDK::Logs::Logger.new(name, version, self)
```

Comment:
@arielvalentin, @robbkidd - Per our discussion during the SIG, I've removed the `@mutex.synchronize` wrapper.

Reply:
We use a "registry" in the `MeterProvider` and `TracerProvider` to ensure the same instance is returned for the identical `name` and `version` pairs. I don't see that requirement in the spec for `LoggerProvider`, and maybe it doesn't matter (e.g. if `Logger` has no state). I'm mildly surprised and need to 🤔 a bit more about it.

Reply:
I have a registry that follows the `MeterProvider` and `TracerProvider` patterns in some code I have queued up for future PRs. I'm happy to bring it into this PR if you'd like to see it now. It became quite necessary once instrumentation was in the mix.

Reply:
Spec says:
> When more than one Logger of the same `name`, `version`, and `schema_url` is created, it is unspecified whether or under which conditions the same or different Logger instances are returned.

So there's some flexibility here. We'll need to implement equality in `Logger` if we choose not to implement a registry.

Reply:
Do we need to fully qualify the class here?
```suggestion
         Logger.new(name, version, self)
```